File: app/src/main/java/com/tcssol/expensetracker/Model/PersonExpViewModel.java
package com.tcssol.expensetracker.Model;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.tcssol.expensetracker.Data.PersonExpDao;
import com.tcssol.expensetracker.Data.PersonExpRepository;

import java.util.List;

public class PersonExpViewModel extends AndroidViewModel {
    public static PersonExpRepository personExpRepository;
    public final LiveData<List<PersonExp>> allExpenses;
    public final LiveData<List<PersonExp>> allExpensesGrouped;

    public PersonExpViewModel(@NonNull Application application) {
        super(application);
        personExpRepository = new PersonExpRepository(application);
        this.allExpenses = personExpRepository.getAllExpenses();
        this.allExpensesGrouped = personExpRepository.getAllExpensesGrouped();
        // ✅ No synchronous DB call here
    }

    public LiveData<List<PersonExp>> getAllExpensesGroupedFiltered(String month, String year) {
        return personExpRepository.getAllExpensesGroupedFiltered(month, year);
    }

    public LiveData<List<PersonExp>> getAllExpensesGrouped() {
        return allExpensesGrouped;
    }

    public LiveData<List<PersonExp>> getAllExpensesFiltered(String month, String year) {
        return personExpRepository.getAllExpensesFiltered(month, year);
    }

    /** Fetches list on a background thread and delivers via callback. */
    public void getAllExpensesList(java.util.function.Consumer<List<PersonExp>> callback) {
        personExpRepository.getAllExpensesList(callback);
    }

    /** Synchronous – only call from a background thread. */
    public List<PersonExp> getAllExpensesListSync() {
        return personExpRepository.getAllExpensesListSync();
    }

    public LiveData<List<PersonExp>> getAllExpenses() { return allExpenses; }

    public static void insert(PersonExp personExp) {
        personExpRepository.insert(personExp);
    }

    public static void delete(PersonExp personExp) { personExpRepository.delete(personExp); }

    public PersonExpDao getPersonExpDao() { return personExpRepository.getPersonExpDao(); }
}

File: app/src/main/java/com/tcssol/expensetracker/Model/PersonExpViewModel.kt
package com.tcssol.expensetracker.Model

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.tcssol.expensetracker.Data.PersonExpDao
import com.tcssol.expensetracker.Data.PersonExpRepository
import com.tcssol.expensetracker.Model.PersonExp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

class PersonExpViewModel(application: Application) : AndroidViewModel(application) {
    val repository: PersonExpRepository = PersonExpRepository(application)
    val allExpenses: Flow<List<PersonExp>> = repository.allExpenses
    val allExpensesGrouped: Flow<List<PersonExp>> = repository.allExpensesGrouped

    fun getAllExpensesGroupedFiltered(month: String, year: String): Flow<List<PersonExp>> {
        return repository.getAllExpensesGroupedFiltered(month, year)
    }

    fun getAllExpensesGrouped(): Flow<List<PersonExp>> {
        return allExpensesGrouped
    }

    fun getAllExpensesFiltered(month: String, year: String): Flow<List<PersonExp>> {
        return repository.getAllExpensesFiltered(month, year)
    }

    fun getAllExpensesList(callback: (List<PersonExp>) -> Unit) {
        viewModelScope.launch(Dispatchers.IO) {
            val list = repository.getAllExpensesListSync()
            callback(list)
        }
    }

    suspend fun getAllExpensesListSync(): List<PersonExp> {
        return repository.getAllExpensesListSync()
    }

    fun getAllExpenses(): Flow<List<PersonExp>> {
        return allExpenses
    }

    fun insert(personExp: PersonExp) {
        viewModelScope.launch(Dispatchers.IO) {
            repository.insert(personExp)
        }
    }

    fun delete(personExp: PersonExp) {
        viewModelScope.launch(Dispatchers.IO) {
            repository.delete(personExp)
        }
    }

    fun getPersonExpDao(): PersonExpDao {
        return repository.getPersonExpDao()
    }
}
